Create optimal_quantization.py

AliceLeBrigant opened the pull request.

johmathe commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,72 @@
+"""
+Optimal quantization
```
Can you describe a little bit more what this deos please?

johmathe commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,72 @@
+S2metric = HypersphereMetric(dimension=2)
+
+TOLERANCE = 1e-5
+IMPLEMENTED = ['S2']
```
use a tuple, not a list

johmathe commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,72 @@
+
+def sample_from(points, size=1):
+    """
+    Sample from the empirical distribution associated to points
```
please be consistent with . at the end of the sentences in the comments.

johmathe commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,72 @@
+    n_points = points.shape[0]
+    dimension = points.shape[-1]
+
+    ind = np.random.randint(low=0, high=n_points, size=size)
```
use full words - ind ?

johmathe commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,72 @@
+    # random initialization of the centers
+    centers = sample_from(points, n_centers)
+
+    gap = 1
```
gap = 1.0 to show that it is a floating point value

johmathe commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,72 @@
+
+    while gap > tolerance:
+        step += 1
+        k = np.floor(step / n_repetition) + 1
```
what's the rationale for starting at step = 1 instead of step = 0?

alebrigant replied:
Starting at step = 0 will replace the center to be updated by the new sample, instead of just moving it in the direction of it (as in any other step > 0). It is equivalent to a slightly different initialization of the centers (only one is different). Since there is no reason for that initialization to be better than the previous one, I start at step = 1.

johmathe replied:
ok you should find a better name for k - hard to understand what this variable mean. Usually k will be used for indexing an array. variables like i,j k should have a trivial behavior like starting at 0 and incrementing. Any other more complex behavior should have a clear name.

alebrigant replied:
Yes Nina had the same comment, so I changed it to ```step_size```.

johmathe commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,72 @@
+
+        centers[ind, :] = new_center
+
+    return centers, step
```
you need unit tests for this file

AliceLeBrigant commented:
Thanks for the reviews ! Here are the last changes:

1) Modifications of optimal_quantization.py: I have added a karcher_mean function that is needed for the test of optimal_quantization. Maybe it would be more coherent to define a class 'DataPoints' of which all the functions defined in optimal_quantization.py could be methods ?

2) Creation of test_optimal_quantization.py: the unit test of optimal quantization consists in verifying that if one looks for only one center, one should find the Karcher mean. However I couldn't guarantee a better precision than 0.5% of the 'diameter' of the set of data points.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+"""
+Optimal quantization of the empirical distribution of a dataset -
```
Nit: why not a point at the end of the line?

alebrigant replied:
I replaced it with a dot.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+
+def diameter_of_data(points, space=None):
+    """
+    Compute the two points that are farthest away from each other in points.
```
NIT: Compute the *distance* between the two points

alebrigant replied:
Yes thanks !

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    return index_closest_neighbor
+
+
+def diameter_of_data(points, space=None):
```
NIT: maybe `diameter` would sound better and add to `riemannian_metric.py`?

alebrigant replied:
Corrected.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    n_points = points.shape[0]
+
+    for i in range(n_points-1):
+        dist_to_neighbors = metric.dist(points[i, :], points[i+1:, :])
```
Doesn't this give only one real number: the distance between point i and point i+1? If so, I don't understand the next line taking the max of a single real number.

alebrigant replied:
No because the second argument contains all the points from i+1 to the end (the ":" is hard to see).

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    """
+    Compute the Karcher mean of points using a Karcher flow algorithm.
+    Return :
+        - the karcher mean
```
NIT: Uppercase letter for *K*archer everywhere.

alebrigant replied:
Corrected.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    while gap > tolerance:
+        tangent_vectors = np.zeros(points.shape)
+
+        for i in range(n_points):
```
You shouldn't need the for loop as geomstats code is "vectorized": `tangent_vectors = metric.log(points, karcher_mean)` should give you what you want.

alebrigant replied:
Thanks, corrected !

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+            tangent_vectors[i, :] = metric.log(points[i, :], karcher_mean)
+
+        tan_vec_update = np.sum(tangent_vectors, axis=0) / n_points
+        karcher_mean_updated = metric.exp(tan_vec_update, karcher_mean)
```
NIT: consistently use "update" or "updated" on L107 and L108?

alebrigant replied:
Karcher_mean is removed.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+        for i in range(n_points):
+            tangent_vectors[i, :] = metric.log(points[i, :], karcher_mean)
+
+        tan_vec_update = np.sum(tangent_vectors, axis=0) / n_points
```
NIT: This is cosmetics, but could you maybe do this in two lines in order to have a better variable name: `tangent_vec_update` maybe?

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    return karcher_mean, step
+
+
+def optimal_quantization(points, n_centers=10, space=None, n_repetition=20,
```
NIT: `n_repetitions` with a 's' since we have `n_centers`.

alebrigant replied:
Corrected.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    """
+    Compute the optimal approximation of points by a smaller number
+    of weighted centers using the Competitive Learning Riemannian
+    Quantization algorithm. Returns :
```
Maybe give a reference for the Competitive Learning Riemannian Quantization algorithm, like a arxiv link?

alebrigant replied:
Good idea, thanks !

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+
+    while gap > tolerance:
+        step += 1
+        k = np.floor(step / n_repetition) + 1
```
Add `import geomstats.backend as gs` at the beginning of the file and use `gs.floor` instead of `np.floor`. Add the corresponding function in `geomstats/backend/numpy.py` if it isn't already. This will allow the code to work with tensorflow and pytorch later.

alebrigant replied:
I have replaced `np` by `gs` everywhere and I have added a few functions to `geomstats/backend/numpy.py`.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+        index = closest_neighbor(sample, centers, space)
+        center_to_update = centers[index, :]
+
+        tan_vec_update = metric.log(sample, center_to_update) / (k+1)
```
NIT: `tangent_vec_update`?

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+        centers[index, :] = new_center
+
+    n_points = points.shape[0]
+    clustering = np.zeros((n_points,))
```
Same: use `gs.zeros`

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+
+    while gap > tolerance:
+        step += 1
+        k = np.floor(step / n_repetition) + 1
```
Could we give another name to `k`? (avoiding single letter variable names).

alebrigant replied:
Corrected.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    gap = 1.0
+    step = 0
+
+    while gap > tolerance:
```
Same as before: introduce a n_max_iterations or n_max_steps maybe?

alebrigant replied:
Corrected.

ninamiolane commented on quantization/plot_quantization_s2.py:
```diff
@@ -0,0 +1,54 @@
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
```
import geomstats.backend as gs and use it instead of np

ninamiolane replied:
geomstats.backend

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+
+    weights = weights / n_points
+
+    return centers, weights, clustering, step
```
Maybe return a dictionary, to keep track of which variable is where in the output tuple?

alebrigant replied:
Good idea, thanks !

ninamiolane commented on quantization/test/test_optimal_quantization.py:
```diff
@@ -0,0 +1,84 @@
+        sample = oq.sample_from(self.points)
+        result = False
+        for i in range(self.n_points):
+            if (self.points[i, :] == sample).all():
```
Could you use `np.allclose(self.points[i, :], sample)`?

alebrigant replied:
Corrected.

ninamiolane commented on quantization/test/test_optimal_quantization.py:
```diff
@@ -0,0 +1,84 @@
+        result = False
+        for i in range(self.n_points):
+            if (self.points[i, :] == sample).all():
+                result = True
```
you can add a `break` when sample is found to be one point, in order to stop the for loop.

alebrigant replied:
Thanks for the tip !

ninamiolane commented on quantization/test/test_optimal_quantization.py:
```diff
@@ -0,0 +1,84 @@
+        closest_neighbor = self.points[index, :]
+        result = False
+        for i in range(self.n_points):
+            if (self.points[i, :] == closest_neighbor).all():
```
Same as above.

ninamiolane commented on quantization/test/test_optimal_quantization.py:
```diff
@@ -0,0 +1,84 @@
+
+import unittest
+
+import numpy as np
```
Use `import geomstats.backend as gs`.

ninamiolane commented on quantization/test/test_optimal_quantization.py:
```diff
@@ -0,0 +1,84 @@
+        for i in range(self.n_points):
+            tangent_vectors[i, :] = self.metric.log(
+                    self.points[i, :], karcher_mean)
+        sum_tan_vec = np.sum(tangent_vectors, axis=0)
```
NIT: `sum_tangent_vecs`?

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    return diameter
+
+
+def karcher_flow(points, space=None, tolerance=TOLERANCE):
```
This looks like: https://github.com/geomstats/geomstats/blob/master/geomstats/riemannian_metric.py#L219

Could we fuse the codes maybe?

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+IMPLEMENTED = ('S1', 'S2')
+
+
+def sample_from(points, size=1):
```
There might be a function doing this already, something like: https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.choice.html
?

alebrigant replied:
It seems like `numpy.random.choice` takes only 1D arrays as entries, and so it cannot be directly used to sample from a set of points in 2 or more dimensions. I can use it to choose an index, but I don't think it would be very different from using `numpy.random.randint` ?

ninamiolane replied:
Ah, true. Too bad, it seems that they wanted to add the option but haven't done it so far: https://github.com/numpy/numpy/pull/7810

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,171 @@
+    return sample
+
+
+def closest_neighbor(point, neighbors, space=None):
```
Add to `riemannian_metric.py`

alebrigant replied:
Done !

AliceLeBrigant commented:
I had missed the `mean` method of `RiemannianMetric`, thanks for pointing it out ! I have removed `karcher_flow` which was indeed doing the same thing, and I have moved `diameter` and `closest_neighbor` to `riemannian_metric.py`.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+IMPLEMENTED = ('S1', 'S2')
+
+
+def sample_from(points, size=1):
```
How about `n_samples=1` instead of `size=1` to be consistent with the other sampling functions?

alebrigant replied:
Corrected.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+    dimension = points.shape[-1]
+
+    index = gs.random.randint(low=0, high=n_points, size=size)
+    sample = points[gs.ix_(index, gs.arange(dimension))]
```
I think I don't understand the `size` parameter. Why not:
``` 
index = gs.random.randint(low=0, high=n_points, size=(n_samples,))
sample = points[index, :]
```
?

alebrigant replied:
It seems that ```size=n_samples``` (previously ```size=size```) gives the same result as ```size=(n_samples,)```.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+        if gs.isclose(gap, 0, atol=tolerance):
+                break
+
+    if iteration is n_max_iterations:
```
This should be `n_max_iterations-1` or the while loop condition above should be <=.

ninamiolane replied:
Use `==` to compare integers.
https://stackoverflow.com/questions/2239737/is-it-better-to-use-is-or-for-number-comparison-in-python

alebrigant replied:
Yes thanks !

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+    else:
+        metric = HypersphereMetric(dimension=2)
+
+    # random initialization of the centers
```
No need for this comment: the code is clear enough and this will save us from meaningless leftover comments if we later change the code but forget to adapt the comments.

alebrigant replied:
Corrected.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+        - a dictionary containing the clusters
+        - the number of steps needed to converge.
+    """
+    if space not in IMPLEMENTED:
```
How about passing `metric` as a parameter of the function and then forget about testing the space? Would this optimal quantization work with any metric?

If so, maybe also add:
`# TODO(alice): Consider adding optimal_quantization as a method of the class RiemannianMetric.`

alebrigant replied:
Yes it would work for any metric and it would indeed make sense to add it as a method of ```RiemannianMetric```. I'll get on it.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+        iteration += 1
+        step_size = gs.floor(iteration / n_repetitions) + 1
+
+        sample = sample_from(points)
```
Could we call this `point` (or something better?) instead of `sample`? 'sample' can mean 'several points' which could be confusing:
https://en.wikipedia.org/wiki/Sample_(statistics)

alebrigant replied:
Corrected.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+        index = metric.closest_neighbor(sample, centers)
+        center_to_update = centers[index, :]
+
+        tangent_vec_update = metric.log(sample, center_to_update) \
```
I think we should take the habit of writing:
`log(point=sample, base_point=center_to_update)`
`exp(tangent_vec=tangent_vec_update, base_point=new_center)`
to make the code clearer and avoid to confuse the order of point wrt base_point and tangent_vec wrt base_point. What do you think?

alebrigant replied:
Yes I agree !

ninamiolane commented on quantization/plot_quantization_s1.py:
```diff
@@ -0,0 +1,31 @@
+"""
+
+import matplotlib.pyplot as plt
+import geomstats.backend as gs
```
imports:
```
import matplotlib.pyplot as plt

import geomstats.backend as gs
import quantization.optimal_quantization as oq

from geomstats.hypersphere import Hypersphere
```

alebrigant replied:
Corrected.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+        if index not in index_list:
+            clusters[index] = list()
+            index_list.append(index)
+        clusters[index].append(i)
```
Is there a reason why we append the index i of the point and not the point itself? Wouldn't be easier to append the point with `for point in points`?

Similarly, do we need to keep track of the centers that are not used? If not, do we need to keep track of the index index of each center, as the key of the dict clusters? 

--> Why not outputing: `clusters, weights, iteration` where clusters is a list of lists of points, weights a list of numbers and iteration is the iteration of convergence?

alebrigant replied:
Yes you're right it is easier to append the point itself, I changed that.

However, even if we want a list of lists, it seems to me that we need to keep track of the unused centers in order to know when to create a new list. In other words, I don't really see how we can organize the clusters other than by numbering the lists, which is the idea behind a dictionary. What do you think ?

alebrigant replied:
Also, I think it is important to keep the ```centers``` as an output, since they provide the "optimal summary" of the data points, in the sense of the Wasserstein distance. What do you think ?

ninamiolane commented on quantization/plot_quantization_s1.py:
```diff
@@ -0,0 +1,31 @@
+    centers, weights, clustering, n_iterations = oq.optimal_quantization(
+                points, n_centers, space='S1', n_repetitions=20, tolerance=1e-6
+                )
+    theta = gs.linspace(0, 2*gs.pi, 100)
```
Could you add this as a class `Circle` in `visualization.py`, similar to the class `Sphere`? Thank you!

https://github.com/geomstats/geomstats/blob/master/geomstats/visualization.py#L64

ninamiolane commented on quantization/plot_quantization_s1.py:
```diff
@@ -0,0 +1,31 @@
+from geomstats.hypersphere import Hypersphere
+
+CIRCLE = Hypersphere(dimension=1)
+
```
Add all the constants at the beginning:
```
N_POINTS = 1000
N_CENTERS = 5
N_REPETITIONS = 20
TOLERANCE=1e-6
```

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,101 @@
+
+def optimal_quantization(points, n_centers=10, space=None, n_repetitions=20,
+                         tolerance=TOLERANCE, n_max_iterations=50000):
+    """
```
Could you add a short explanation about how you use `n_repetitions`? thanks.

ninamiolane commented on quantization/plot_quantization_s2.py:
```diff
@@ -0,0 +1,54 @@
+    plt.figure(1)
+    ax = plt.subplot(111, projection="3d", aspect="equal")
+    color = np.random.rand(n_centers, 3)
+    ax.plot_wireframe(sphere.sphere_x,
```
Use methods of the sphere: `sphere.draw`, `sphere.add_points`, `sphere.draw_points`, etc
https://github.com/geomstats/geomstats/blob/master/geomstats/visualization.py#L94

alebrigant replied:
I have tried to use them in the best way possible, however it would be more satisfactory if the sphere was not drawn as many times as the number of clusters (I need each cluster to be drawn in a different color and so I repeat ```sphere.draw``` ```n_clusters``` times). Maybe this could be fixed by creating a new list of points each time ```add_points``` is called, that could be plotted in a different color (or some other way) ?

Also, I would recommend to change the ```alpha=0.5``` in ```draw``` to a lower value such as ```alpha=0.2``` so that the points are more visible, or to make it into a adjustable parameter.

alebrigant replied:
Never mind, the sphere is not drawn when using ```sphere.draw_points``` instead of ```sphere.draw```. Sorry about that !

ninamiolane commented on quantization/plot_quantization_s1.py:
```diff
@@ -0,0 +1,30 @@
+                n_repetitions=N_REPETITIONS, tolerance=TOLERANCE
+                )
+    visualization.plot(centers, space='S1', color='red')
+
```
Does it make sense to also draw the points, each in the color corresponding to its center, to mimic plot_quantization_s2 example?

alebrigant replied:
Yes ! I added that.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,98 @@
+    return sample
+
+
+def optimal_quantization(points, metric, n_centers=10, n_repetitions=20,
```
Put 
```
N_CENTERS = 10
N_REPETITIONS = 20
N_MAX_ITERATIONS = 5000
```
at the beginning of the file.

ninamiolane commented on quantization/plot_quantization_s2.py:
```diff
@@ -0,0 +1,50 @@
+    color = gs.random.rand(N_CENTERS, 3)
+    for i in range(N_CENTERS):
+        cluster_i = gs.vstack([point for point in clusters[i]])
+        sphere = visualization.Sphere()
```
Do we need a new Sphere for each cluster?

alebrigant replied:
I changed visualization.py in that direction. See https://github.com/geomstats/geomstats/pull/141.

ninamiolane commented on quantization/plot_quantization_s2.py:
```diff
@@ -0,0 +1,50 @@
+        cluster_i = gs.vstack([point for point in clusters[i]])
+        sphere = visualization.Sphere()
+        sphere.add_points(cluster_i)
+        if i == 0:
```
Do we need this edge case? If so, could we tackle it directly in visualization.py?

alebrigant replied:
Same as above, see PR #141.

ninamiolane commented on quantization/optimal_quantization.py:
```diff
@@ -0,0 +1,98 @@
+    Return :
+        - n_centers centers
+        - n_centers weights between 0 and 1
+        - a dictionary containing the clusters
```
Add something like: "where each key is the cluster index, and its value is the lists of points belonging to the cluster."?

ninamiolane commented:
Thanks! Let's first agree on the visualisation.py version and I'll have another look at this one after.
